subsys/bluetooth/host/l2cap.c

```diff
@@ -50,6 +50,12 @@
 #define L2CAP_CONN_TIMEOUT	K_SECONDS(40)
 #define L2CAP_DISC_TIMEOUT	K_SECONDS(2)
 
+/* Dedicated pool for disconnect buffers so they are guaranteed to be send
+ * even in case of data congestion due to flooding.
+ */
+NET_BUF_POOL_FIXED_DEFINE(disc_pool, 1,
+			  BT_L2CAP_BUF_SIZE(CONFIG_BT_L2CAP_TX_MTU), NULL);
```

Comment:
Does this really guarantee allocation success? Remember that the bt_conn_tx allocation is something that can also fail if any sending is done from the sys wq.

Reply:
Now that we are using RX thread for fixed channels this might not be necessary since we block with K_FOREVER.

Reply:
I guess it depends where the disconnect request is sent from. What context is that, and are there multiple possibilities? Only the server-side is run through the RX thread. Is that what you were trying to fix? I suppose the client side is completely application-dependent, i.e. what context a disconnection is requested from.

Reply:
Right, I was thinking just about the signalling but you are right the application can call it in any context, including on syswq.

Reply:
Yes, it does because the flooding still causes the credits to fail so we have to shutdown and disconnect.